src/include/btree.i

```diff
@@ -35,6 +35,31 @@ __wt_page_is_modified(WT_PAGE *page)
 #define	WT_ALLOC_OVERHEAD	32U
 
 /*
+ * Track a field in the cache. Use atomic CAS so that we can reliably avoid
+ * decrementing the cache below zero - since we use an unsigned number.
+ * Track if we would go below zero in a diagnostic build - something has gone
+ * wrong.
+ */
+#ifdef	HAVE_DIAGNOSTIC
+#define WT_CACHE_DECR(session, f, sz) do {				\
+	uint64_t __val = f;						\
+	uint64_t __sz = WT_MIN(__val, sz);				\
+	if (__sz < sz)							\
+		__wt_errx(session, "%s underflow: decrementing %"	\
+		    WT_SIZET_FMT, #f, sz);				\
+	while (!WT_ATOMIC_CAS8(f, __val, __val - __sz))			\
+		__val = f, __sz = WT_MIN(__val, __sz);			\
+} while (0)
+#else
+#define WT_CACHE_DECR(session, f, sz) do {				\
+	uint64_t __val = f;						\
+	uint64_t __sz = WT_MIN(__val, sz);				\
+	while (!WT_ATOMIC_CAS8(f, __val, __val - __sz))			\
+		__val = f, __sz = WT_MIN(__val, __sz);			\
+} while (0)
+#endif
+
+/*
```

Comment:
@agorrod, I like using CAS better than the "back it out if it underflows" change I used in 23eac61d89dae1f875287f888ceb833b88b89b5f.

I don't like repeated error messages (once we start down this path, we're unlikely to recover).

I also don't see the point in repeatedly decrementing? Why would that make the value any more correct? We could just set the value to 0 (that's the same as what's there now, unless an increment races with us in which case we could use a "future" increment to help satisfy our "current" decrement, which just means we'll hit the same problem when that "future" increment tries to do its decrement.